### DISCUSSION TITLE
TidHTTP.Get and Stream Size

### DISCUSSION THREAD

**Discussion by tothpaul** [IndySockets/Indy#592] [Q&A]
Hello,

I've made a OpenStack Swift (for Hubic) client with TidHTTP, when I call 
```
// download a file up to 5Gb
  idHTTP.Get('https://.../v1/{accout}/{container}/{object}', FileStream)
```
the Stream is resized according to the HTTP response Content-Length.

but sometime, I have a read timout and the file is incomplet, the stream is already truncated to the downloaded size...but with Swith I can add a CustomHeader like that
 
```
// resume download
  idHTTP.CustomHeaders.Add('Range: bytes=' + IntToStr(FileStream.Size) + '-' + IntToStr(ExpectedSize));
  idHTTP.Get('https://.../v1/{accout}/{container}/{object}', FileStream)
```
in that case the HTTP header returns the size of portion added, and the FileStream should concat data at its end.

Is there a way to do that without a custom stream or a temp file ? 

```
// resume download
  idHTTP.CustomHeaders.Add('Range: bytes=' + IntToStr(FileStream.Size) + '-' + IntToStr(ExpectedSize));
  idHTTP.Get('https://.../v1/{accout}/{container}/{object}', Temp);
  FileStream.CopyFrom(Temp);
```

OpenStack Swifth API is available there
https://developer.openstack.org/api-ref/object-store/?expanded=get-object-content-and-metadata-detail

**Comment by rlebeau** [comment]
First off, you should be using the TIdHTTP.Request.Ranges property instead:
```
with idHTTP.Request.Ranges.Add do
begin
  StartPos := FileStream.Size;
  EndPos := ExpectedSize;
end;
```  
Or:
```
idHTTP.Request.Ranges.Add.Text := Format('%d-%d', [FileStream.Size, ExpectedSize]);
```
Or:
```
idHTTP.Request.Ranges.Text := Format('bytes=%d-%d', [FileStream.Size, ExpectedSize]);
```

That being said, TIdHTTP takes the target TStream's current Position into account.  Simply set the FileStream.Position to the desired start Position, and that is where TIdHTTP will write new data at.  So, in your case, you should be able to set the FileStream.Position to the FileStream.Size before calling TIdHTTP.Get(), eg:
```
FileStream.Position := FileStream.Size;
FilePos := FileStream.Position;
if (FilePos = 0) or (FilePos < ExpectedSize) then
begin
  if Pos > 0 then 
  begin
    with idHTTP.Request.Ranges.Add do
    begin
      StartPos := FilePos;
      EndPos := ExpectedSize;
    end;
  end;
  idHTTP.Get('https://.../v1/{accout}/{container}/{object}', FileStream);
end;
```
Indy even has a TIdAppendFileStream class that presets the Position to the Size for you when it opens a file:
```
FileStream := TIdAppendFileStream.Create('filename');
FilePos := FileStream.Position;
// same as above ...
```

**Comment by tothpaul** [comment]
great ! thanks
